mpeg2/src/interleaving_muxer.rs

```diff
@@ -0,0 +1,463 @@
+use super::EncodeError;
+use crate::muxer::{Muxer, Packet, Stream, StreamConfig};
+use alloc::vec::Vec;
+use core2::io::Write;
+use std::collections::VecDeque;
+use std::time::Duration;
+
+const TS_33BIT_MASK: u64 = 0x1FFFFFFFF;
+const TS_ROLLOVER_THREAD: u64 = 1 << 20;
+
+pub struct InterleavingMuxer<W: Write> {
+    inner: Muxer<W>,
+    max_buffer_duration_90khz: u64,
+    largest_ts_in_buffer: u64,
+    streams: Vec<InterleavingStream>,
+}
+
+pub struct InterleavingStream {
+    inner: Stream,
+    buffered_packets: VecDeque<Packet>,
+    last_written_ts: u64,
+}
+
+impl<W: Write> InterleavingMuxer<W> {
+    pub fn new(w: W, max_buffer_duration: Duration) -> Self {
+        let muxer = Muxer::new(w);
+        Self {
+            inner: muxer,
+            max_buffer_duration_90khz: (max_buffer_duration.as_millis() * 90) as u64,
+            largest_ts_in_buffer: 0,
+            streams: vec![],
+        }
+    }
+
+    pub fn add_stream(&mut self, config: StreamConfig) {
+        let inner = self.inner.new_stream(config);
+        let interleaving_stream = InterleavingStream {
+            inner,
+            buffered_packets: VecDeque::with_capacity(6),
+            last_written_ts: 0,
+        };
+        self.streams.push(interleaving_stream);
+    }
+
+    #[inline]
+    fn get_most_significant_bit(&self, mut n: u64) -> u64 {
+        while n > 1 && (n & (n - 1) != 0) {
+            n &= n - 1;
+        }
+        n
+    }
```

Comment:
No need for `&self` here. And there are standard library functions that can be used here to avoid the loop and use intrinsics instead:

```suggestion
    #[inline]
    fn get_most_significant_bit(n: u64) -> u64 {
        (1 << (64 - n.leading_zeros())) >> 1
    }
```

Reply:
👍  I didn't know there is such a handy method in rust!